master/master.go

```diff
@@ -5,54 +5,121 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ec2"
 )
 
+type Master struct {
+	svc              *ec2.EC2 // safe to be used concurrently
+	instances        map[string]slave
+	jobs             map[int64]*job
+	jobsChan         chan lib.Job
+	heartbeatChan    chan lib.Heartbeat
+	statusChan       chan chan statusSummary
+	newTasks         []*lib.Task
+	scheduledTasks   []*lib.Task
+	scheduleTicker   *time.Ticker // channel to instruct the main loop to schedule tasks
+	controllerTicker *time.Ticker
+	controller       controller
+	taskSize         int64
+	quit             chan bool
+}
+
 type slave struct {
 	tasks    []*lib.Task
-	maxTasks int
-	// TODO others
+	maxSlots int
+	instance *ec2.Instance // can't populate this easily
 }
 
-type Master struct {
-	instances      map[string]slave
-	jobs           map[int]*job
-	jobsChan       chan lib.Job
-	heartbeatChan  chan lib.Heartbeat
-	statusChan     chan chan string // dummy
-	newTasks       []*lib.Task
-	scheduledTasks []*lib.Task
-	controllerChan chan string // dummy
-	scheduleChan   chan bool   // channel to instruct the main loop to schedule tasks
+type scheduler interface {
+	schedule(map[string]slave) string
 }
 
+type controller struct {
+	dt       time.Duration
+	kp       float64
+	kd       float64
+	ki       float64
+	prevErr  float64
+	integral float64
+}
+
+type statusSummary struct {
+	instances map[string]slave // TODO *ec2.Instance probably won't serialise
+	jobs      map[int64]*job
+}
+
+// Init creates the master object
 func Init() Master {
-	// TODO initialise Master correctly
-	return Master{}
+	// set some defaults
+	return Master{
+		svc:              nil, // initialised in Run
+		instances:        make(map[string]slave),
+		jobs:             make(map[int64]*job),
+		jobsChan:         make(chan lib.Job),
+		heartbeatChan:    make(chan lib.Heartbeat),
+		statusChan:       make(chan chan statusSummary),
+		newTasks:         make([]*lib.Task, 0),
+		scheduledTasks:   make([]*lib.Task, 0),
+		scheduleTicker:   nil, // initialised in Run
+		controllerTicker: nil, //initialised in Run
+		controller: controller{
+			dt:       time.Minute * 2,
+			kp:       1,
+			kd:       0,
+			ki:       0,
+			prevErr:  0,
+			integral: 0,
+		},
+		taskSize: 6400 * 1000 * 1000,
+		quit:     make(chan bool),
+	}
 }
 
+// Run starts the master
 func (m *Master) Run() {
-	http.HandleFunc(lib.JobsCreatePath, m.jobsHandler)
-	http.HandleFunc(lib.HeartbeatPath, m.heartbeatHandler)
-	http.HandleFunc(lib.StatusPath, m.statusHandler)
+	// initialise the nils
+	m.initAWS()
+	m.controllerTicker = time.NewTicker(m.controller.dt)
+	// TODO test how this performs when a lot of tasks get submitted.
+	m.scheduleTicker = time.NewTicker(time.Duration(100/(len(m.newTasks)+1)) * time.Millisecond)
 
-	go http.ListenAndServe(lib.Port, nil)
+	// setup and run http
+	http.HandleFunc(lib.JobsCreatePath, makeJobsHandler(m.jobsChan))
```

Comment:
Yeah, I initially kept it outside because the handlers took pointer receiver, but they don't anymore so they can be moved inside now.